### PR TITLE
CNV-32097: New list page filter!!! 

### PR DESCRIPTION
**Pull request by upalatucci**
## 📝 Description

In the future, we'll be able to use the proxy pod for filtering and add other search filters like 'Search by IP' 

Edit: Removed the column management feature. We'll add it in the next pr

## 🎥 Demo

**Video**

https://github.com/kubevirt-ui/kubevirt-plugin/assets/29160323/89b3892b-b34b-4891-919b-c267643d1d0c

**Before**
Screenshot from 2023-09-27 16-00-59
Screenshot from 2023-09-27 16-00-56
Screenshot from 2023-09-27 16-00-53
Screenshot from 2023-09-27 16-00-49

**After**
Screenshot from 2023-09-27 15-57-56
Screenshot from 2023-09-27 15-57-53
Screenshot from 2023-09-27 15-57-48
Screenshot from 2023-09-27 15-57-42

### CODE REVIEW

File: src/utils/components/ColumnManagementModal/constants.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+export const MAX_VIEW_COLS = 9;
+
+export const NAME_COLUMN_ID = 'name';
+export const READ_ONLY_COLUMNS = new Set([NAME_COLUMN_ID]);
```
Comment (metalice) [inline src/utils/components/ColumnManagementModal/constants.ts:4]: if its a set and not a string const , name should be as any var ur are using in the app

**Comment by metalice** [comment]
@upalatucci could u sperate the PR into 2 prs? 1 for filtering and another for column management?

File: src/utils/components/ListPageFilter/AutocompleteInput.tsx [hunk 2]
```diff
@@ -0,0 +1,86 @@
+import React, { Dispatch, FC, SetStateAction, useEffect, useState } from 'react';
+import classNames from 'classnames';
+
+import SearchFilter from './SearchFilter';
+import { fuzzyCaseInsensitive, labelParser } from './utils';
+
+const MAX_SUGGESTIONS = 5;
```
Comment (metalice) [inline src/utils/components/ListPageFilter/AutocompleteInput.tsx:7]: can move to the util file?

File: src/utils/components/ListPageFilter/AutocompleteInput.tsx [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      <span className={className}>{suggestion}</span>
+    </button>
+  );
+};
+
+type SuggestionLineProps = {
+  className?: string;
+  onClick: (param: string) => void;
+  suggestion: string;
+};
```
Comment (metalice) [inline src/utils/components/ListPageFilter/AutocompleteInput.tsx:84]: what do u think to move to a new component?

File: src/utils/components/ListPageFilter/AutocompleteInput.tsx [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  className,
+  data,
+  onSuggestionSelect,
+  placeholder,
+  setTextValue,
+  showSuggestions,
+  textValue,
+}) => {
+  const [suggestions, setSuggestions] = useState<string[]>();
+  const [visible, setVisible] = useState(true);
```
Comment (metalice) [inline src/utils/components/ListPageFilter/AutocompleteInput.tsx:19]: < boolean >

File: src/utils/components/ListPageFilter/ListPageFilter.tsx [hunk 5]
```diff
@@ -0,0 +1,296 @@
+import React, { FC, useCallback, useState } from 'react';
```
Comment (metalice) [inline src/utils/components/ListPageFilter/ListPageFilter.tsx:1]: please split this file as its 300 lines

File: src/utils/components/ListPageFilter/ListPageFilter.tsx [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+  const { t } = useKubevirtTranslation();
+
+  const [isDropdownOpen, setIsDropdownOpen] = useState(false);
+  const [isSelectOpen, setIsSelectOpen] = useState(false);
+
+  // Generate rowFilter items and counts. Memoize to minimize re-renders.
+  const generatedRowFilters = useDeepCompareMemoize(generateRowFilters(rowFilters ?? [], data));
+
+  // Reduce generatedRowFilters once and memoize
+  const [filters, filtersNameMap, filterKeys] = React.useMemo<
```
Comment (metalice) [inline src/utils/components/ListPageFilter/ListPageFilter.tsx:53]: please remove React. from this file where needed

File: src/utils/components/ListPageFilter/ListPageFilter.tsx [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+    filterKeys,
+    filters,
+  });
+
+  const textFilters = useSearchFiltersParameters();
+
+  const [searchType, setSearchType] = useState<keyof typeof STATIC_SEARCH_FILTERS>(
+    STATIC_SEARCH_FILTERS.name,
+  );
+  const [searchInputText, setSearchInputText] = useState(
```
Comment (metalice) [inline src/utils/components/ListPageFilter/ListPageFilter.tsx:69]: < string? >

File: src/utils/components/ListPageFilter/ListPageFilter.tsx [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+    searchType !== STATIC_SEARCH_FILTERS.labels ? textFilters[searchType] : '',
+  );
+
+  const onToggleClick = () => {
```
Comment (metalice) [inline src/utils/components/ListPageFilter/ListPageFilter.tsx:73]: there is a need for this func? or simply use setDrop.. where needed?

File: src/utils/components/ListPageFilter/ListPageFilter.tsx [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+    setIsDropdownOpen(!isDropdownOpen);
+  };
+
+  const onSelect = (event, value) => {
+    setSearchInputText(value !== STATIC_SEARCH_FILTERS.labels ? textFilters[value] : '');
```
Comment (metalice) [inline src/utils/components/ListPageFilter/ListPageFilter.tsx:78]: searchType !== STATIC_SEARCH_FILTERS.labels ? textFilters[searchType] : '' is already in the init sate. so maybe a func for it?

File: src/utils/components/ListPageFilter/ListPageFilter.tsx [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+    setSearchType(value);
+    setIsDropdownOpen(false);
+  };
+
+  const applyFilters = useCallback(
```
Comment (metalice) [inline src/utils/components/ListPageFilter/ListPageFilter.tsx:83]: callback is needed here?

File: src/utils/components/ListPageFilter/ListPageFilter.tsx [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+  const onRowFilterSelect = (event) => {
+    updateRowFilterSelected([event?.target?.id]);
+  };
+
+  const clearAll = () => {
+    updateRowFilterSelected(selectedRowFilters);
+    applyTextFilters(STATIC_SEARCH_FILTERS.name);
+    applyTextFilters(STATIC_SEARCH_FILTERS.labels);
+    setSearchInputText('');
+  };
```
Comment (metalice) [inline src/utils/components/ListPageFilter/ListPageFilter.tsx:130]: can be in util?

File: src/utils/components/ListPageFilter/SearchFilter.tsx [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+type SearchFilterProps = {
+  className?: string;
+  label?: string;
+  placeholder?: string;
+} & TextInputProps;
+
+const SearchFilter: FC<SearchFilterProps> = (props) => {
+  const { className, label, placeholder, ...otherInputProps } = props;
+  const { t } = useKubevirtTranslation();
+  const placeholderText = placeholder ?? t('Filter {{label}}...', { label });
```
Comment (metalice) [inline src/utils/components/ListPageFilter/SearchFilter.tsx:16]: if label is null?

Comment (metalice) [reply src/utils/components/ListPageFilter/SearchFilter.tsx:16]: label is not reuired

File: src/utils/components/ListPageFilter/hooks.ts [hunk 13]
```diff
@@ -0,0 +1,87 @@
+import { useCallback, useMemo } from 'react';
```
Comment (metalice) [inline src/utils/components/ListPageFilter/hooks.ts:1]: why all hooks together?

File: src/utils/components/ListPageFilter/hooks.ts [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+}) => [string[], (selected: string[]) => void];
+
+export const useRowFiltersParameters: UseRowFiltersParametersType = ({ filterKeys, filters }) => {
+  const location = useLocation();
+  const { setOrRemoveQueryArgument } = useQueryParamsMethods();
+  const queryParams = useMemo(() => new URLSearchParams(location.search), [location]);
+
+  const selectedRowFilters = useMemo(
+    () =>
+      Object.values(filterKeys)
```
Comment (metalice) [inline src/utils/components/ListPageFilter/hooks.ts:16]: u think better add .? of Object.XX failes we are getting a crash

Comment (metalice) [reply src/utils/components/ListPageFilter/hooks.ts:16]: for map & flat

File: src/utils/components/ListPageFilter/hooks.ts [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+        .map((values) => queryParams.get(values)?.split(',') ?? [])
+        .flat(),
+    [queryParams, filterKeys],
+  );
+
+  const syncRowFilterParams = useCallback(
+    (selected: string[]) => {
+      Object.entries(filters).forEach(([key, value]) => {
```
Comment (metalice) [inline src/utils/components/ListPageFilter/hooks.ts:24]: same as above

File: src/utils/components/ListPageFilter/hooks.ts [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+
+    Object.entries(newParams).forEach(([k, v]) => {
+      if (params.get(k) !== v) {
+        update = true;
+        params.set(k, v);
+      }
+    });
+
+    if (update) {
+      const url = new URL(window.location.href);
```
Comment (metalice) [inline src/utils/components/ListPageFilter/hooks.ts:63]: why simply not use the new ones? why checking?

File: src/utils/components/ListPageFilter/hooks.ts [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+    const params = new URLSearchParams(window.location.search);
+    let update = false;
+    keys.forEach((k) => {
+      if (params.has(k)) {
+        update = true;
+        params.delete(k);
+      }
+    });
+    if (update) {
+      const url = new URL(window.location.href);
```
Comment (metalice) [inline src/utils/components/ListPageFilter/hooks.ts:78]: same as aboev

File: src/views/virtualmachines/list/VirtualMachinesList.tsx [hunk 18]
```diff
@@ -130,7 +129,7 @@ const VirtualMachinesList: FC<VirtualMachinesListProps> = ({ kind, namespace })
     }));
   };
 
-  const [columns, activeColumns] = useVirtualMachineColumns(namespace, pagination, data);
+  const [_, activeColumns] = useVirtualMachineColumns(namespace, pagination, data);
```
Comment (metalice) [inline src/views/virtualmachines/list/VirtualMachinesList.tsx:132]: why nit needed anymore and only active?

File: src/utils/components/ListPageFilter/utils.ts [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+  const requirements = [];
+  const matchLabels = isOldFormat(selector) ? selector : selector.matchLabels;
+  const { matchExpressions } = selector;
+  Object.keys(matchLabels || {})
+    .sort()
+    .forEach((k) => {
+      requirements.push(createEquals(k, matchLabels[k]));
+    });
+  (matchExpressions || []).forEach((me) => {
+    requirements.push(me);
```
Comment (metalice) [inline src/utils/components/ListPageFilter/utils.ts:134]: const a = Object.keys(matchLabels || {})
    .sort().map(...createEquals(k, matchLabels[k]))

req.. = [...a, ...match ]

File: src/utils/components/ListPageFilter/SearchFilter.tsx (hunk 12 above)
Comment (upalatucci) [reply src/utils/components/ListPageFilter/SearchFilter.tsx:16]: i'll make it required

File: src/utils/components/ListPageFilter/hooks.ts (hunk 14 above)
Comment (upalatucci) [reply src/utils/components/ListPageFilter/hooks.ts:16]: i'll add `|| {}` so `selectedRowFilters` is always defined

File: src/utils/components/ListPageFilter/hooks.ts (hunk 17 above)
Comment (upalatucci) [reply src/utils/components/ListPageFilter/hooks.ts:78]: in that case keys is at least one or empty array

Comment (upalatucci) [reply src/utils/components/ListPageFilter/hooks.ts:78]: Not in the `setAllQueryArguments` so i'll fix there

File: src/utils/components/ListPageFilter/utils.ts [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+  if (requirement.operator === 'NotIn') {
+    return `${requirement.key} notin (${toArray(requirement.values).join(',')})`;
+  }
+  if (requirement.operator === 'GreaterThan') {
+    return `${requirement.key} > ${requirement.values[0]}`;
+  }
+  if (requirement.operator === 'LessThan') {
+    return `${requirement.key} < ${requirement.values[0]}`;
+  }
+  return '';
```
[the remark this answers is not in the document]
Comment (upalatucci) [reply src/utils/components/ListPageFilter/utils.ts:113]: Yes it was a copy paste from the SDK
I'll refactor

File: src/views/virtualmachines/list/VirtualMachinesList.tsx (hunk 18 above)
Comment (upalatucci) [reply src/views/virtualmachines/list/VirtualMachinesList.tsx:132]: unused variable. Eslint has a problem with that. 
When we'll add the columnManagementModal in the ListPageFilter, we'll use it back again.

File: src/utils/components/ListPageFilter/constants.ts [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+  labels: 'Labels',
+  name: 'Name',
+} as const;
+
+// t('Search by labels...')
+// t('Search by name...')
+export const STATIC_SEARCH_FILTERS_PLACEHOLDERS = {
+  labels: 'Search by labels...',
+  name: 'Search by name...',
+} as const;
```
[the remark this answers is not in the document]
Comment (upalatucci) [reply src/utils/components/ListPageFilter/constants.ts:18]: it's an alternative to enums. It says to typescript that this object is not modified by javascript and suggests the entries while typing.

File: src/utils/components/ListPageFilter/hooks.ts (hunk 16 above)
Comment (upalatucci) [reply src/utils/components/ListPageFilter/hooks.ts:63]: I think it means that if you want to remove something that was already not applied, do not update

File: src/views/virtualmachines/list/VirtualMachinesList.tsx (hunk 18 above)
Comment (metalice) [reply src/views/virtualmachines/list/VirtualMachinesList.tsx:132]: yes i know ive added the linter :) 
I wanted to know why it isn't in use, u answered :)

File: src/utils/components/ListPageFilter/utils.ts [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+  const s = new Set(b);
+  return a.filter((x) => s.has(x));
+};
+
+export const getLabelsAsString = (obj: K8sResourceCommon): string[] => {
+  const requirements = toRequirements(obj.metadata.labels);
+  return Object.values(requirements).map(requirementToString);
+};
+
+export const labelParser = (resources: any[]): Set<string> => {
```
[the remark this answers is not in the document]
Comment (upalatucci) [reply src/utils/components/ListPageFilter/utils.ts:83]: yes with `K8sResourceCommon`

[the remark this answers is not in the document]
Comment (upalatucci) [reply src/utils/components/ListPageFilter/utils.ts]: fixed

File: src/utils/components/ListPageFilter/utils.ts [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+  return a.filter((x) => s.has(x));
+};
+
+export const getLabelsAsString = (obj: K8sResourceCommon): string[] => {
+  const requirements = toRequirements(obj.metadata.labels);
+  return Object.values(requirements).map(requirementToString);
+};
+
+export const labelParser = (resources: K8sResourceCommon[]): Set<string> => {
+  return resources.reduce((acc: Set<string>, resource: any) => {
```
Comment (metalice) [inline src/utils/components/ListPageFilter/utils.ts:84]: any

**Review by avivtur — COMMENTED** [review]
Great work :smile: :+1: 
Added some comments nothing too big

File: src/utils/components/ListPageFilter/AutocompleteInput.tsx [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+import classNames from 'classnames';
+
+import { MAX_SUGGESTIONS } from './constants';
+import SearchFilter from './SearchFilter';
+import SuggestionLine from './SuggestionLine';
+import { fuzzyCaseInsensitive, labelParser } from './utils';
+
+type AutocompleteInputProps = {
+  className?: string;
+  data?: any;
```
[the remark this answers is not in the document]
Comment (upalatucci) [reply src/utils/components/ListPageFilter/AutocompleteInput.tsx:11]: i think `K8sResourceCommon` should work

File: src/utils/components/ListPageFilter/AutocompleteInput.tsx [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+  useEffect(() => {
+    if (textValue && showSuggestions) {
+      const processed = labelParser(data);
+      // User input without whitespace
+      const processedText = textValue.trim().replace(/\s*=\s*/, '=');
+      const filtered = [...processed]
+        .filter((item) => fuzzyCaseInsensitive(processedText, item))
+        .slice(0, MAX_SUGGESTIONS);
+      setSuggestions(filtered);
+    }
```
[the remark this answers is not in the document]
Comment (upalatucci) [reply src/utils/components/ListPageFilter/AutocompleteInput.tsx:51]: Yes right. Good idea. I'll drop the showSuggestions props for this reason.

File: src/utils/components/ListPageFilter/SearchFilter.tsx [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+type SearchFilterProps = {
+  className?: string;
+  placeholder: string;
+} & TextInputProps;
+
+const SearchFilter: FC<SearchFilterProps> = (props) => {
+  const { className, placeholder, ...otherInputProps } = props;
+
+  return (
+    <div className="has-feedback">
```
[the remark this answers is not in the document]
Comment (upalatucci) [reply src/utils/components/ListPageFilter/SearchFilter.tsx:15]: No. Breaks the input width. This is how it was done in the SDK

File: src/utils/components/ListPageFilter/utils.ts (hunk 23 above)
Comment (upalatucci) [reply src/utils/components/ListPageFilter/utils.ts]: solved

File: src/utils/components/ListPageFilter/RowFilters.tsx [hunk 27]
```diff
@@ -0,0 +1,117 @@
...
+  };
+
+  if (rowFilters.length === 0) return null;
+  return (
+    <ToolbarItem>
+      {Object.keys(filters).reduce(
+        (acc, key) => (
+          <ToolbarFilter
+            chipGroupCollapsedText={t('{{numRemaining}} more', {
+              numRemaining: '${remaining}',
```
[the remark this answers is not in the document]
Comment (avivtur) [reply src/utils/components/ListPageFilter/RowFilters.tsx:53]: Cool thanks for the explanation :)

**Pull request merged**